File: src/schemas/hotels.py
from pydantic import BaseModel, Field, ConfigDict
# Создаем схемы данных, их типы, чтобы потом использовать в ручках, пайдентик автоматически будет проверять тип данных,
# который мы передаем. И чтобы fastapi в ручке мог принять такой класс, наследуемся от базовой модели
from typing import List, Literal

class HotelOwnerCreate(BaseModel):
    title: str
    location: str
    images: list[str] = Field(default_factory=list)
    location_ru: str | None = None
    title_ru: str | None = None

class HotelAdd(HotelOwnerCreate):
    owner_id: int | None = None
    status: Literal["draft", "published", "archived"] = "draft"

class Hotel(HotelAdd):
    id : int
    images: List[str] = []
    model_config = ConfigDict(from_attributes=True) # приводим ответ Алхимии к виду словаря, чтобы пайдентик мог с ним работать


class HotelPatch(BaseModel):
    title: str | None = Field(None)
    location: str | None = Field(None)
    images: list[str] | None = None
    location_ru: str | None = None
    title_ru: str | None = None

class HotelStatusPatch(BaseModel):
    status: Literal["draft", "published", "archived"]

File: hotels-frontend/src/api/hotels.ts
import { api } from "./client";

export interface Hotel {
  id: number;
  title: string;
  location: string;
  location_ru?: string;
  title_ru?: string;
  images?: string[];
  owner_id?: number | null;
  status?: "draft" | "published" | "archived";
}

export async function getHotels(params?: {
  page?: number;
  per_page?: number;
  title?: string;
  location?: string;
  images?: string[];
}) {
  const res = await api.get<Hotel[]>("/hotels", { params });
  return res.data;
}

export async function createHotel(payload: {
  title: string;
  location: string;
  images?: string[];
}) {
  const res = await api.post("/hotels", payload);
  return res.data;
}

export async function deleteHotel(id: number) {
  const res = await api.delete(`/hotels/${id}`);
  return res.data;
}

export async function updateHotel(
  id: number,
  payload: { title?: string; location?: string; images?: string[] }
) {
  const res = await api.patch(`/hotels/${id}`, payload);
  return res.data;
}

export async function createOwnerHotel(payload: {
  title: string;
  location: string;
  images?: string[];
}) {
  const res = await api.post("/hotels/owner", payload);
  return res.data;
}

export async function getMyHotels() {
  const res = await api.get<Hotel[]>("/hotels/owner/my");
  return res.data;
}

export async function patchMyHotel(
  id: number,
  payload: { title?: string; location?: string; images?: string[] }
) {
  const res = await api.patch(`/hotels/owner/${id}`, payload);
  return res.data;
}

export async function publishMyHotel(id: number) {
  const res = await api.post(`/hotels/owner/${id}/publish`);
  return res.data;
}

export async function unpublishMyHotel(id: number) {
  const res = await api.post(`/hotels/owner/${id}/unpublish`);
  return res.data;
}